### PR TITLE
close binary-proto-lookup connection after connection-lifetime

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

[PulsarClient](https://github.com/yahoo/pulsar/blob/master/pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java#L121) creates a separate dedicated connection with broker for a given service-lookup url. So, in case of broker's cold-restart, all clients may end-up creating lookup-connection to same one broker and this connection will be persisted for entire lifetime of the client, which may create lookup-overload on that one broker. So, lookup connection should not be persistent and it should be closed after sometime.

### Modifications

- disconnect lookup connection after x seconds (default = 10 min)

### Result

Lookup connection will not be persistent and it will avoid one broker to be overloaded with lookup requests.

### CODE REVIEW

**Comment by merlimat** [comment]
> So, in case of broker's cold-restart, all clients may end-up creating lookup-connection to same one broker 

Uhm, all the lookup from a _single_ client will end up on the same broker/discovery-service instance,  but there is no reason why lookups from multiple clients should do the same.

**Comment by rdhabalia** [comment]
>but there is no reason why lookups from multiple clients should do the same.

In case of cold-restart as soon as first broker comes up all the client tries to do lookup and they will endup creating lookup-connection with one broker. Also, lookup connection is persistent so, it will be used forever for that client so, that one broker will endup serving all the lookup requests. So, does it make sense to not keep lookup-connection persistent and can be closed after fixed time..??

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java [hunk 1]
```diff
@@ -111,22 +116,36 @@ public void initChannel(SocketChannel ch) throws Exception {
     private static final Random random = new Random();
 
     public CompletableFuture<ClientCnx> getConnection(final InetSocketAddress address) {
+        return getConnection(address, -1);
+    }
+
+    /**
+     * 
+     * @param address
+     *            remote client {@link InetSocketAddress} of the to connect
```
Comment (saandrews) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java]: sentence seems to be incomplete?

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ClientCnx.java [hunk 2]
```diff
@@ -377,6 +378,7 @@ SocketAddress serverAddrees() {
         ctx.writeAndFlush(cmd).addListener(writeFuture -> {
             if (!writeFuture.isSuccess()) {
                 log.warn("{} Failed to send request to broker: {}", ctx.channel(), writeFuture.cause().getMessage());
+                pendingRequests.remove(requestId);
```
Comment (saandrews) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ClientCnx.java]: Shouldn't it be getAndRemovePendingLookupRequest?

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java [hunk 3]
```diff
@@ -170,6 +189,15 @@ public void initChannel(SocketChannel ch) throws Exception {
...
             });
+
+            // close the connection after connection-lifetime completes
+            if (connectionLifetimeInSecond > 0) {
+                scheduler.schedule(() -> {
+                    if (cnx != null && cnx.channel().isActive()) {
+                        cnx.channel().disconnect();
+                    }
+                }, connectionLifetimeInSecond, TimeUnit.SECONDS);
+            }
```
Comment (saandrews) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java]: Are we closing the connection irrespective of it's active or not? We should close it only if it's inactive for the last n minutes.

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ClientCnx.java (hunk 2 above)
Comment (rdhabalia) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ClientCnx.java]: no, actually we had a bug where if we failed to publish message on socket then we should remove from the queue.  line:341 removes from `lookup-PendingRequests` and this one removes from `pendingRequests`

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java (hunk 3 above)
Comment (rdhabalia) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java]: I think our goal is to not keep lookup connection persistent forever and close after n minute. however, we can add check to close connection if not active for last n minute. i will add another commit.

Comment (rdhabalia) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java]: @saandrews updated PR with the change.

**Comment by merlimat** [comment]
> In case of cold-restart as soon as first broker comes up all the client tries to do lookup and they will endup creating lookup-connection with one broker. 

Good point.. In that case all the connections would stick to same broker.

My only concern is that by closing the connection after X time will make every lookup after a failover to open a new TCP connection (one per client) and re-do authentication, while if we maintain the connection open it would be ready to use. I'm not sure that would be a big deal though.

**Comment by saandrews** [comment]
In that case, it would behave similar to our http lookup for the first request. Should be ok I think.

**Comment by saandrews** [comment]
There are few more scenarios to address: Handling failure in case broker is unable to serve the lookup(similar to http 500) and closing it only if inactive for the configured duration. Will be addressed in a separate PR.

**Comment by merlimat** [comment]
> In that case, it would behave similar to our http lookup for the first request. Should be ok I think.

True, it would kind of be the same. One other option could be to react on the lookups being rejected by the broker (from #181) and close the connection only if errors are received in a certain timeframe.

**Comment by rdhabalia** [comment]
>One other option could be to react on the lookups being rejected by the broker (from #181) and close the connection only if errors are received in a certain timeframe.

Does it mean client should close the connection when broker gives `TooManyRequestError` back? It means
1. If client doesn't receive `TooManyRequestError` then client will keep this persistent-connection forever. So, there is a still chance that all clients go to the same broker and can still create load on one broker if traffic is < 20K (`maxConcurrentLookupRequest=20000`)

But I think I agree with the idea that if client gets `TooManyRequestError` in certain timeframe then client should disconnect the connection and try to connect to different broker (as that broker is already overloaded).

Same as like #265 we close http-connection on internal-server-error, we need to do it for binary-lookup. So, I will combine 2 changes in new PR:
1. close binary-proto-lookup connection on 500 error
2. close connection if client receives `TooManyLookupRequestError` in certain timeframe.

And we can think of if we don't need this PR-change (close persistent-connection) then we may close this one.

Any thought?

**Comment by merlimat** [comment]
My reasoning is: 
 * All client will end up in a single broker only if it's the only one left in cluster
 * At that point the broker should be overwhelmed by lookups, if it is not... then we should be good anyway
 * If we force close the connections, we can give the clients a chance to reconnect to the other brokers that are being brought up

> 1. close binary-proto-lookup connection on 500 error
> 2. close connection if client receives TooManyLookupRequestError in certain timeframe.

Yep, that's what I was thinking

**Comment by merlimat** [comment]
@rdhabalia Should this PR be closed in favor of #274 ?

**Comment by rdhabalia** [comment]
yes, let's close this one.

**Pull request closed without merging**